### PR TITLE
Initial support for Metal Performance Shaders (MPS)

### PR DESCRIPTION
**Pull request by dameikle**
Makes adjustments to provide initial support for Apple Silicon Metal Performance Shaders (MPS) to provide a base for further optimisation.

This has been tested on basic training and inference on both M1 and M3 Macs. It has also been used for inference of larger production grade models that have been trained on CUDA.

output

### CODE REVIEW

File: eole/modules/multi_headed_attn.py [hunk 1]
```diff
@@ -460,12 +460,27 @@ def _compute_attention(
                         query,
                         key,
                         value,
-                        ~mask if mask is not None else None,
+                        ~mask if (mask is not None and not causal) else None,
```
Comment (vince62s) [inline eole/modules/multi_headed_attn.py:463]: why are you adding "not causal" as a condition ?

File: eole/modules/multi_headed_attn.py [hunk 2]
```diff
@@ -460,12 +460,27 @@ def _compute_attention(
...
                         self.dropout_p,
                         is_causal=causal,
                     )
             attn = None
-
+        elif (
```
Comment (vince62s) [inline eole/modules/multi_headed_attn.py:468]: why can't we use the same code above ?

File: eole/train_single.py [hunk 3]
```diff
@@ -137,6 +138,7 @@ def main(config, device_id):
     # config is the full TrainConfig namespace here
 
     configure_process(config, device_id)
+    print("** Device ID " + str(device_id))
```
Comment (vince62s) [inline eole/train_single.py:141]: remove print

File: eole/utils/misc.py [hunk 4]
```diff
@@ -74,6 +74,51 @@ def use_gpu(config):
...
+    if device_type == "auto":
+        device_type = get_device_type()
+
+    if device_type == "cuda":
+        return torch.cuda.amp.autocast()
+    elif device_type == "mps":
+        return torch.amp.autocast(device_type="mps")
+    elif device_type == "cpu":
+        return torch.cpu.amp.autocast()
+    else:
```
Comment (vince62s) [inline eole/utils/misc.py:118]: is this possible?

File: eole/modules/multi_headed_attn.py (hunk 1 above)
Comment (dameikle) [reply eole/modules/multi_headed_attn.py:463]: It should technically throw an error when is_causal and an attention mask is provided:
https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html

Base Python implementation does this, the CUDA backends don't.

File: eole/modules/multi_headed_attn.py (hunk 2 above)
Comment (dameikle) [reply eole/modules/multi_headed_attn.py:468]: We definately could in the current case, I was splitting it out so we could add further MPS specifics later.

File: eole/utils/misc.py (hunk 4 above)
Comment (dameikle) [reply eole/utils/misc.py:118]: No, on reflection, shouldn't be. Will always have CPU.

File: eole/modules/multi_headed_attn.py (hunk 1 above)
Comment (vince62s) [reply eole/modules/multi_headed_attn.py:463]: I know but I can't see how this can be right. At training time, on the decoder side, it is causal=True, still we have the mask.

**Comment by francoishernandez** [comment]
Thanks a lot @dameikle ! Quite curious to see what we can achieve on Apple silicon chips.
I'll try and test this some time this week.

File: eole/modules/multi_headed_attn.py (hunk 1 above)
Comment (francoishernandez) [reply eole/modules/multi_headed_attn.py:463]: Stumbled upon this issue some time ago as well. I agree this is not clear.
Maybe we could make the condition on `causal` instead of `mask` to make things cleaner.
I don't have all the background of this module but I guess the point of supporting `is_causal` here is just in case we don't have a `mask`. (Not sure this case exists for now, but why not supporting it as long it's not hurting anything.)

**Comment by francoishernandez** [comment]
@dameikle, I managed to run this fine on an M3 Macbook Air 👌

A few questions:
1. What kind of environment do you use on your mac(s)? I had a few issues setting things up (especially with some requirements not liking python3.12, had to setup python3.10.) If you have any tips to make things easier for other users, it might be good to create a new entry somewhere in the docs (e.g. in the FAQ sections).
2. Did you have to make some adaptations to the recipe(s) you tested? Seems like some of the bash commands are not all supported on default macOS (or my shell is wonky, possible).
3. You mention "inference of larger production grade models". Could you provide some details on this? E.g. a minimal benchmark of model size / performance (speed, memory, etc.) would very useful information, especially if we plan on "further optimisation"!

File: eole/modules/multi_headed_attn.py (hunk 1 above)
Comment (dameikle) [reply eole/modules/multi_headed_attn.py:463]: I had originally went with the the check on _causal_ as you propose @francoishernandez. I personally think it's the most clear.

**Comment by dameikle** [comment]
> What kind of environment do you use on your mac(s)? I had a few issues setting things up (especially with some requirements not liking python3.12, had to setup python3.10.) If you have any tips to make things easier for other users, it might be good to create a new entry somewhere in the docs (e.g. in the FAQ sections).

I normally use Virtual Envs created from Homebrew installed Python to avoid clashes with the macOS installed version. I can put together some documentation.

> Did you have to make some adaptations to the recipe(s) you tested? Seems like some of the bash commands are not all supported on default macOS (or my shell is wonky, possible).

I was mainly doing things by hand as well as using others models I'd trained elsewhere, so wasn't trying them out to be honest. Can have a wee look at them and see if we can make them work across both.

> You mention "inference of larger production grade models". Could you provide some details on this? E.g. a minimal benchmark of model size / performance (speed, memory, etc.) would very useful information, especially if we plan on "further optimisation"!

Sounds like a plan. Will pull information together.

File: eole/modules/multi_headed_attn.py (hunk 1 above)
Comment (vince62s) [reply eole/modules/multi_headed_attn.py:463]: for sake of clarity let's make the code look like:
with sdpa_kernel([SDPBackend.MATH, SDPBackend.EFFICIENT_ATTENTION]):
                    attn_output = scaled_dot_product_attention(
                        query,
                        key,
                        value,
                        ~mask if mask is not None else None,
                        self.dropout_p,
                        is_causal=False,
                    )
I know is_causal is False by default in sdpa but at least it keeps things clearer.
The only case Model is "causal" is at training on decoder side for self-attn only (not context) and since our code passes a mask (sequential + triangular) it is ok to use the mask parameter in sdpa.

File: eole/modules/multi_headed_attn.py (hunk 2 above)
Comment (vince62s) [reply eole/modules/multi_headed_attn.py:468]: For now I would prefer to keep the code as compact as possible until you add further MPS stuff later on. Thanks.

**Comment by francoishernandez** [comment]
Thanks for the updates @dameikle!
Can we merge or do you have pending topics to be pushed?
(Docs can be updated separately when you'll find time to put it together.)

**Comment by dameikle** [comment]
@francoishernandez - I think it's good to merge now. I was starting to work on the docs and taking a look at the recipies last night but like you say, can do that in seperate PR.
